### ISSUE TITLE
Usability suggestions

### ISSUE THREAD

**Issue by Ostrzyciel** [Kaszanas/SC2InfoExtractorGo#52]
Some usability suggestions based on a few minutes with the project:

- I'd suggest to only use lowercase names for directories, which is the standard on Linux. Upper/lowercase doesn't matter on Windows, but it does matter on Linux, creating additional headaches for multi-platform software (it's why Steam changed to lowercase dirs a long time ago). It's easiest to just keep all paths lowercase.
- GNU-style long options (de facto standard) have `--` as their prefix, not `-`. I'd suggest to use the standard style if possible. The only thing that uses the other that comes to my mind is the `java` executable, which is a weird relict of its own time. As a side note: the undocumented `--help` option works with the current version of the program.
- The tool now outputs nothing when running it without arguments. This is a bit confusing, maybe display sth like "nothing to process"?
- Currently the Docker images are built only on release. I'd suggest to build the images on every commit to the main and dev branches, to allow for continuous testing of the build (does it still work?). Otherwise, the CI pipeline may turn out not to work at all during a release. It also helps with using dev builds of the tool – they are not only for developers, but also for people who want to use the latest code. These people don't have to know how to use a go compiler (I sure as hell don't). Don't worry, the CI is free ;)

**Comment by Kaszanas** [comment]
> I'd suggest to only use lowercase names for directories, which is the standard on Linux. Upper/lowercase doesn't matter on Windows, but it does matter on Linux, creating additional headaches for multi-platform software (it's why Steam changed to lowercase dirs a long time ago). It's easiest to just keep all paths lowercase.

I agree with this, the only issue currently was that the overall structure of this project is not up to my standards. Why? Some directories contain Golang modules while others serve only the purpose of being a data input etc.

By default I would love to have a `src` directory and have the Golang code reside there fully. It is unreasonable for me to have `main.go` in the root directory with `README`, licenses, and other documentation-like files.

> GNU-style long options (de facto standard) have -- as their prefix, not -. I'd suggest to use the standard style if possible. The only thing that uses the other that comes to my mind is the java executable, which is a weird relict of its own time. As a side note: the undocumented --help option works with the current version of the program.

 Golang `"flags"` package unfortunately does not allow for that. I hold the same view.

> The tool now outputs nothing when running it without arguments. This is a bit confusing, maybe display sth like "nothing to process"?

This is because originally the defaults were set. I think then that there is no way to check if the user passed anything in that case. Alternative:
- Delete some defaults like input and output directory, and display help if these were not provided. 
- Have the ENTRYPOINT calling `-help` which exits the program after outputting the usage information:

```go
	flag.Parse()

	if *help {
		flag.Usage()
		os.Exit(1)
	}
```

> Currently the Docker images are built only on release. I'd suggest to build the images on every commit to the main and dev branches, to allow for continuous testing of the build (does it still work?). Otherwise, the CI pipeline may turn out not to work at all during a release. It also helps with using dev builds of the tool – they are not only for developers, but also for people who want to use the latest code. These people don't have to know how to use a go compiler (I sure as hell don't). Don't worry, the CI is free ;)

Fully agree, that is why on CI runs on any pull request (I have added the development dockerfile build to the CI):

```
      # Check if the main dockerfile builds:
      - name: Build "Production" Docker Image
        run: |
          make build

        # Check if the main dockerfile builds:
      - name: Build "Development" Docker Image
        run: |
          make build_dev

      # TODO: Running containerized tests:
      # - name: Run Docker Tests With Compose
      #   run: |
      #     make action_compose_test
```

So the flow is as follows (1) write some code, (2) open PR, (3) see if it builds and runs on test data, (4) merge.

**Comment by Ostrzyciel** [comment]
As for the CI workflow – CI builds in PRs are useful, yes, but continuous builds of the merged code in the dev branch are useful as well. That way, when you merge any code it's automatically built and pushed to the Docker registry, allowing users to pull bleeding-edge Docker images right away. This is something that I use often in other projects.

**Comment by Kaszanas** [comment]
> As for the CI workflow – CI builds in PRs are useful, yes, but continuous builds of the merged code in the dev branch are useful as well. That way, when you merge any code it's automatically built and pushed to the Docker registry, allowing users to pull bleeding-edge Docker images right away. This is something that I use often in other projects.

The only questions then are:
- What should be the tagging scheme?

For example:
main branch - `sc2infoextractorgo:main`?
dev branch - `sc2infoextractorgo:dev`?
releases - either `sc2infoextractorgo:latest` for the most recent release tag, or `sc2infoextractorgo:sem.ver.tag` explicit tagging with semver?

What about distributing the development dockerfile that can be used as a VS Code devcontainer?

**Comment by Ostrzyciel** [comment]
The tagging scheme you are proposing makes sense :)
As for releases, I'd have both `latest` and semver.

Tests – would make sense. The test results on the PR branch will not necessarily be the same as on the target branch (it's a merge, after all). It doesn't cost anything, so I see no reason not to do it.

A devcontainer sounds like a good idea – though personally I have (almost) never used them, so I don't have any specific recommendations.

**Comment by Kaszanas** [comment]
> The tagging scheme you are proposing makes sense :) As for releases, I'd have both `latest` and semver.
> 
> Tests – would make sense. The test results on the PR branch will not necessarily be the same as on the target branch (it's a merge, after all). It doesn't cost anything, so I see no reason not to do it.
> 
> A devcontainer sounds like a good idea – though personally I have (almost) never used them, so I don't have any specific recommendations.

Then I seem to know everything that I need to go forward with that. I will be hosting data for tests in a separate DocherHub repository so it can be easily mounted (then the code for downloading test data will be unecessary, so that there is less code to maintain). After that I should be able to fix some of the issues that were raised in:
- https://github.com/Kaszanas/SC2InfoExtractorGo/issues/55
- https://github.com/Kaszanas/SC2InfoExtractorGo/issues/50
- https://github.com/Kaszanas/SC2InfoExtractorGo/issues/51

And after all that is done and a PR to main is initialized, this should solve:
- https://github.com/Kaszanas/SC2InfoExtractorGo/issues/54

**Comment by Kaszanas** [comment]
Most of this is done at this point. The Docker Image release upon new pushes to specific branches will be addressed under #76 

I am closing this.
